File: DVSM/coupling_audit.py
# SPDX-License-Identifier: AGPL-3.0-only
"""
coupling_audit.py — the FORBIDDEN-FEEDBACK-COUPLING firewall for DVSM, built on weltwerk/verify/residual_channel.

The DVSM deployment manifest (ReadMe.rs §5) forbids specific couplings and names "Observer Contamination —
diagnostics begin influencing state evolution" as a failure mode. Today those guarantees are enforced only
by Rust's borrow checker (no *syntactic* write) and by inspection. This module enforces them EMPIRICALLY: a
diagnostic channel X must carry NO information about a future dynamics channel Y beyond the legitimate
drivers Z. The honest discriminator is the confounder-conditioned mutual information:

    I(X ; Y_next | Z_legit)  ≈ null   ⇒  AIR_GAP_HELD          (the borrow-checker promise is also informational)
    I(X ; Y_next | Z_legit)  > null   ⇒  a residual leak; stress it with a candidate confounder W:
        stable under (Z, W)           ⇒  OBSERVER_CONTAMINATION (a real forbidden coupling)
        dissolves under (Z, W)        ⇒  CONFOUNDED_ARTIFACT    (explained by a missing confounder, not a leak)

`borrow-checker-clean ≠ air-gap-sound`: the type system proves no syntactic write; the CMI audit probes for an
informational leak the types cannot see. `residual-CMI ≠ channel` until mis-specification-stable.

IDENTIFIABILITY BOUNDARY (load-bearing, recorded honestly): a coupling is detectable from telemetry only if
the diagnostic X has variance INDEPENDENT of the conditioning set Z. `Ω→V` and `ν→λ` are identifiable (Ω is a
slow independent accumulation; λ has no legitimate driver, so Z is empty). A diagnostic that is itself a
near-deterministic FUNCTION of legit state (e.g. stiffness ≈ 2|z₀|) is UNIDENTIFIABLE by conditioning —
"stiffness→z" is indistinguishable from "z→z". `undetected ≠ absent`; the firewall states where it is blind.

Every verdict routes through the shared honesty contract (AnalysisResult) and registers a graded Claim.
"""
from __future__ import annotations

import os
import sys
from dataclasses import dataclass
from typing import Callable, List, Optional, Sequence, Tuple

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "weltwerk", "verify"))
from residual_channel import audit, ResidualChannelResult          # noqa: E402
from artifacts import AnalysisResult, Finding, Limitation           # noqa: E402
from claim_ledger import Claim                                      # noqa: E402

from dvsm_reference import StepRecord                               # noqa: E402

K = 3            # bins per channel
REPS = 60        # shuffle-null reps per audit


# ---- quantile binning (each channel → ~K equal-mass symbols) -------------------------------------
def _quantile_edges(values: Sequence[float], k: int) -> List[float]:
    xs = sorted(values)
    if not xs:
        return []
    return [xs[min(len(xs) - 1, int(len(xs) * q / k))] for q in range(1, k)]


def _binner(values: Sequence[float], k: int) -> Callable[[float], int]:
    edges = _quantile_edges(values, k)

    def f(v: float) -> int:
        lo = 0
        for e in edges:
            if v <= e:
                return lo
            lo += 1
        return lo
    return f


# ---- coupling specification ----------------------------------------------------------------------
@dataclass(frozen=True)
class CouplingSpec:
    """One forbidden coupling. X = the diagnostic (source); Y = the future dynamics (target); Z = the
    LEGITIMATE determinants of Y; W = a candidate confounder for the mis-specification stress."""
    name: str
    manifest_rule: str
    x: Callable[[StepRecord], float]
    y: Callable[[StepRecord], float]
    z: Callable[[StepRecord], Tuple[float, ...]]
    w: Callable[[StepRecord], Tuple[float, ...]]
    identifiable: bool = True
    note: str = ""


# the manifest's forbidden couplings (ReadMe.rs §5 + dvsm_one_file "Ω isolated witness, no backfeed")
COUPLINGS: Tuple[CouplingSpec, ...] = (
    CouplingSpec(
        "omega_to_v", "NO Ω → V (long-term drift cannot influence instantaneous velocity)",
        x=lambda r: r.omega0, y=lambda r: r.v0_next,
        z=lambda r: (r.drive, r.v0), w=lambda r: (r.s0,),
        identifiable=True),
    CouplingSpec(
        "novelty_to_lambda", "NO ν → λ (novelty spikes cannot modulate global dissipation)",
        x=lambda r: r.novelty, y=lambda r: r.lambda_eff,
        z=lambda r: (0.0,), w=lambda r: (r.z_energy,),
        identifiable=True),
    CouplingSpec(
        "stiffness_to_z", "NO Stiffness → Dynamics (V17-K outputs are diagnostic-only)",
        x=lambda r: r.stiffness, y=lambda r: r.z0_next,
        z=lambda r: (r.drive, r.z0), w=lambda r: (r.s0,),
        identifiable=False,
        note="stiffness ≈ 2|z₀| is a near-function of the conditioned z₀ ⇒ unidentifiable by conditioning."),
)
_BY_NAME = {c.name: c for c in COUPLINGS}

_VERDICT = {
    "CONSISTENT_WITH_NULL": "AIR_GAP_HELD",
    "RESIDUAL_MISSPEC_STABLE": "OBSERVER_CONTAMINATION",
    "RESIDUAL_MISSPEC_FRAGILE": "CONFOUNDED_ARTIFACT",
    "RESIDUAL_DEPENDENCE": "SUSPECT_UNSTRESSED",
}
# UNIDENTIFIABLE: for a coupling whose diagnostic is a near-deterministic function of the conditioned legit
# state, a positive CMI cannot be told apart from binning-resolution confounding. The firewall DECLINES to
# rule rather than emit a false OBSERVER_CONTAMINATION. This is an epistemic state (not measured), not a
# verdict of clean. `detected-on-unidentifiable ≠ contamination`.


@dataclass(frozen=True)
class CouplingResult:
    name: str
    manifest_rule: str
    verdict: str                 # AIR_GAP_HELD | OBSERVER_CONTAMINATION | CONFOUNDED_ARTIFACT | UNIDENTIFIABLE | SUSPECT_UNSTRESSED
    result: ResidualChannelResult
    identifiable: bool
    note: str

    def as_analysis(self) -> AnalysisResult:
        r = self.result
        findings = (
            Finding("MANIFEST_RULE", "forbidden-coupling", self.manifest_rule),
            Finding("RESIDUAL_CMI", "forbidden-coupling",
                    f"I(X;Y_next|Z)={r.cmi:.4f} vs null {r.null_mean:.4f}±{r.null_std:.4f} (z={r.z_score:.1f})"),
            Finding("VERDICT", "forbidden-coupling", self.verdict),
        )
        lims = [
            Limitation("forbidden-coupling", "a verdict is about THIS trace + the MODELED legit set Z; "
                                             "residual-CMI ≠ channel until mis-specification-stable"),
            Limitation("forbidden-coupling", "AIR_GAP_HELD is absence-of-evidence at this window/conditioning, "
                                             "not proof of no coupling. proves-the-procedure ≠ proves-the-phenomenon"),
        ]
        if not self.identifiable:
            lims.append(Limitation("identifiability", self.note + " undetected ≠ absent"))
        return AnalysisResult(source_trace=(), scope="forbidden-coupling",
                              findings=findings, limitations=tuple(lims))

    def as_claim(self) -> Claim:
        r = self.result
        if self.verdict == "OBSERVER_CONTAMINATION":
            return Claim(f"COUP::{self.name}",
                         f"the forbidden coupling '{self.name}' carries residual information into dynamics.",
                         "MEASURED", f"I(X;Y_next|Z)={r.cmi:.3f} > null, stable under (Z,W).",
                         "a precise mechanism or magnitude in the Rust kernel — only that the reference trace leaks.",
                         "the residual dissolves under a further candidate confounder or finer windowing.")
        if self.verdict == "CONFOUNDED_ARTIFACT":
            return Claim(f"COUP::{self.name}",
                         f"an apparent '{self.name}' leak dissolves under (Z,W): a missing-confounder artifact.",
                         "MEASURED", "I(X;Y_next|Z) elevated but ≈ null under (Z,W).",
                         "that the air-gap holds — only that THIS residual is confounder-explained.",
                         "the residual survives every candidate confounder ⇒ promote to contamination.")
        if self.verdict == "UNIDENTIFIABLE":
            return Claim(f"COUP::{self.name}",
                         f"'{self.name}' is unidentifiable from telemetry: the diagnostic is a function of the "
                         f"conditioned legit state.", "NOT_MEASURED", self.note or "diagnostic ⊂ conditioning set.",
                         "presence OR absence of the coupling — the method cannot separate leak from confounding.",
                         "an instrument that varies the diagnostic independently of the legit state.")
        if self.verdict == "AIR_GAP_HELD":
            return Claim(f"COUP::{self.name}",
                         f"no residual dependence for '{self.name}' beyond the legitimate drivers Z.",
                         "MEASURED", f"I(X;Y_next|Z)={r.cmi:.3f} ≈ shuffle null.",
                         "absence of the coupling in the Rust kernel — only that none is visible here.",
                         "a residual appears under finer windowing, a richer Z, or the real kernel trace.")
        return Claim(f"COUP::{self.name}",
                     f"'{self.name}' shows residual dependence but was not mis-specification-stressed.",
                     "UNDERDETERMINED", "residual detected; no misspec_fn ran.",
                     "whether the residual is a channel or an artifact.",
                     "running the (Z,W) stress resolves it.")


# ---- the audit -----------------------------------------------------------------------------------
def audit_coupling(trace: List[StepRecord], spec: CouplingSpec, *, k: int = K, reps: int = REPS,
                   seed: int = 0) -> CouplingResult:
    """Test whether diagnostic X leaks into future dynamics Y beyond legitimate drivers Z (stress with W)."""
    xs_raw = [spec.x(r) for r in trace]
    ys_raw = [spec.y(r) for r in trace]
    z_dims = list(zip(*[spec.z(r) for r in trace]))          # tuple of per-dim series
    w_dims = list(zip(*[spec.w(r) for r in trace]))

    bx, by = _binner(xs_raw, k), _binner(ys_raw, k)
    z_binners = [_binner(d, k) for d in z_dims]
    w_binners = [_binner(d, k) for d in w_dims]

    xs = [bx(v) for v in xs_raw]
    ys = [by(v) for v in ys_raw]
    zs = [tuple(b(d[i]) for b, d in zip(z_binners, z_dims)) for i in range(len(trace))]
    ws = [tuple(b(d[i]) for b, d in zip(w_binners, w_dims)) for i in range(len(trace))]

    samples = list(zip(xs, ys, zs))

    def refine_w(_s):                                        # re-condition on (Z, W): the decisive stress
        return [(xs[i], ys[i], (zs[i], ws[i])) for i in range(len(trace))]

    r = audit(samples, reps=reps, seed=seed, misspec_fns=(refine_w,))
    verdict = _VERDICT.get(r.decision, "SUSPECT_UNSTRESSED")
    if not spec.identifiable:
        verdict = "UNIDENTIFIABLE"          # decline to rule — the method cannot separate leak from confounding
    return CouplingResult(spec.name, spec.manifest_rule, verdict, r, spec.identifiable, spec.note)


def audit_all(trace: List[StepRecord], *, k: int = K, reps: int = REPS, seed: int = 0) -> List[CouplingResult]:
    """Audit every manifest coupling — side by side, NO fused scalar. The plurality IS the finding."""
    return [audit_coupling(trace, c, k=k, reps=reps, seed=seed) for c in COUPLINGS]


# ---- planted-case validator (proves the PROCEDURE, on identifiable couplings) ---------------------
def planted_validator(*, n: int = 6000, reps: int = REPS) -> dict:
    """For each identifiable coupling: a clean trace must read AIR_GAP_HELD and a trace with that coupling
    planted must read OBSERVER_CONTAMINATION. Sound iff the procedure separates them."""
    from dvsm_reference import gen_clean, gen_contaminated
    out = {}
    for c in COUPLINGS:
        if not c.identifiable:
            continue
        clean = audit_coupling(gen_clean(n, seed=1), c, reps=reps, seed=11)
        dirty = audit_coupling(gen_contaminated(c.name, n, seed=2), c, reps=reps, seed=22)
        out[c.name] = {
            "clean_verdict": clean.verdict, "clean_cmi": clean.result.cmi,
            "dirty_verdict": dirty.verdict, "dirty_cmi": dirty.result.cmi,
            "separates": clean.verdict == "AIR_GAP_HELD" and dirty.verdict == "OBSERVER_CONTAMINATION",
        }
    return out


def main():
    print("coupling_audit.py — forbidden-feedback-coupling firewall (borrow-checker-clean ≠ air-gap-sound)\n")
    v = planted_validator()
    for name, r in v.items():
        print(f"  {name:20s} clean={r['clean_verdict']:22s}(CMI={r['clean_cmi']:.3f})  "
              f"planted={r['dirty_verdict']:22s}(CMI={r['dirty_cmi']:.3f})  separates={r['separates']}")
    print("\n  a CHANNEL/OBSERVER_CONTAMINATION verdict = a diagnostic leaks into dynamics; AIR_GAP_HELD is")
    print("  absence-of-evidence, not proof. residual-CMI ≠ channel; proves-the-procedure ≠ proves-the-phenomenon.")


if __name__ == "__main__":
    main()

File: Rust/src/coupling_audit.rs
// SPDX-License-Identifier: AGPL-3.0-only
//! The forbidden-coupling taxonomy — a Rust port of the decision layer in `DVSM/coupling_audit.py`, built on
//! top of the validated [`crate::residual_channel`] core. It asks whether a diagnostic channel X leaks
//! information into a future dynamics channel Y beyond the legitimate drivers Z, and stresses any residual
//! with a candidate confounder W:
//!
//!   I(X;Y_next | Z)  ~ null            ⇒  AIR_GAP_HELD
//!   I(X;Y_next | Z)  > null, stable/(Z,W) ⇒  OBSERVER_CONTAMINATION  (a real forbidden coupling)
//!   I(X;Y_next | Z)  > null, fragile/(Z,W) ⇒  CONFOUNDED_ARTIFACT     (a missing-confounder artifact)
//!
//! IDENTIFIABILITY BOUNDARY: if the diagnostic X is itself a near-deterministic FUNCTION of the conditioned
//! legit state Z, a positive CMI cannot be told apart from binning-resolution confounding. The firewall then
//! DECLINES to rule (UNIDENTIFIABLE) rather than emit a false contamination. `undetected ≠ absent`;
//! `detected-on-unidentifiable ≠ contamination`; `borrow-checker-clean ≠ air-gap-sound`.
//!
//! Continuous channels are quantile-binned (the same scheme as the Python `_binner`). The verdicts are
//! validated on planted, deterministic regimes in `tests/coupling.rs`, cross-checked against the Python
//! reference.

use crate::artifacts::{AnalysisResult, Finding, Limitation};
use crate::residual_channel::{audit, ChannelDecision, Sample};

const K: usize = 3; // bins per channel

/// Quantile binner: each channel → ~K equal-mass symbols. Faithful port of `_quantile_edges`/`_binner`.
pub struct Binner {
    edges: Vec<f64>,
}

impl Binner {
    pub fn new(values: &[f64], k: usize) -> Self {
        let mut xs = values.to_vec();
        xs.sort_by(|a, b| a.partial_cmp(b).expect("no NaN in telemetry channel"));
        let n = xs.len();
        let mut edges = Vec::new();
        if n > 0 {
            for q in 1..k {
                let idx = ((n * q) / k).min(n - 1); // = int(len*q/k) for positive ints
                edges.push(xs[idx]);
            }
        }
        Binner { edges }
    }

    pub fn bin(&self, v: f64) -> i64 {
        let mut lo = 0i64;
        for &e in &self.edges {
            if v <= e {
                return lo;
            }
            lo += 1;
        }
        lo
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CouplingVerdict {
    AirGapHeld,
    ObserverContamination,
    ConfoundedArtifact,
    SuspectUnstressed,
    Unidentifiable,
}

impl CouplingVerdict {
    pub fn as_str(&self) -> &'static str {
        match self {
            CouplingVerdict::AirGapHeld => "AIR_GAP_HELD",
            CouplingVerdict::ObserverContamination => "OBSERVER_CONTAMINATION",
            CouplingVerdict::ConfoundedArtifact => "CONFOUNDED_ARTIFACT",
            CouplingVerdict::SuspectUnstressed => "SUSPECT_UNSTRESSED",
            CouplingVerdict::Unidentifiable => "UNIDENTIFIABLE",
        }
    }
}

/// One coupling to audit. X = diagnostic (source); Y = future dynamics (target); Z = legitimate determinants
/// of Y (each a continuous series); W = candidate confounder(s) for the mis-specification stress.
#[derive(Debug, Clone)]
pub struct CouplingInput {
    pub name: String,
    pub manifest_rule: String,
    pub note: String,
    pub x: Vec<f64>,
    pub y: Vec<f64>,
    pub z_dims: Vec<Vec<f64>>,
    pub w_dims: Vec<Vec<f64>>,
    pub identifiable: bool,
    pub reps: usize,
    pub seed: u64,
}

#[derive(Debug, Clone)]
pub struct CouplingResult {
    pub name: String,
    pub manifest_rule: String,
    pub verdict: CouplingVerdict,
    pub mi: f64,
    pub cmi: f64,
    pub null_mean: f64,
    pub z_score: f64,
    pub identifiable: bool,
    pub note: String,
}

impl CouplingResult {
    pub fn as_analysis(&self) -> AnalysisResult {
        let mut lims = vec![
            Limitation::new(
                "forbidden-coupling",
                "a verdict is about THIS trace + the MODELED legit set Z; residual-CMI != channel until \
                 mis-specification-stable",
            ),
            Limitation::new(
                "forbidden-coupling",
                "AIR_GAP_HELD is absence-of-evidence at this window/conditioning, not proof of no coupling. \
                 proves-the-procedure != proves-the-phenomenon",
            ),
        ];
        if !self.identifiable {
            lims.push(Limitation::new(
                "identifiability",
                format!("{} undetected != absent", self.note),
            ));
        }
        AnalysisResult::new(
            "forbidden-coupling",
            vec![
                Finding::new("MANIFEST_RULE", "forbidden-coupling", self.manifest_rule.clone()),
                Finding::new(
                    "RESIDUAL_CMI",
                    "forbidden-coupling",
                    format!("I(X;Y_next|Z)={:.4} vs null {:.4} (z={:.1})", self.cmi, self.null_mean, self.z_score),
                ),
                Finding::new("VERDICT", "forbidden-coupling", self.verdict.as_str()),
            ],
            lims,
        )
        .expect("coupling as_analysis always satisfies the honesty contract")
    }
}

fn join_bins(binners: &[Binner], dims: &[Vec<f64>], i: usize) -> String {
    let parts: Vec<String> = binners.iter().zip(dims).map(|(b, d)| b.bin(d[i]).to_string()).collect();
    parts.join(",")
}

/// Audit one coupling: bin the channels, build (X, Y, Z) samples, define the (Z, W) re-conditioning, run the
/// residual-channel audit, and map its decision to a coupling verdict (declining as UNIDENTIFIABLE when the
/// coupling is not identifiable from telemetry).
pub fn audit_coupling(inp: &CouplingInput) -> CouplingResult {
    let n = inp.x.len();
    let bx = Binner::new(&inp.x, K);
    let by = Binner::new(&inp.y, K);
    let zbs: Vec<Binner> = inp.z_dims.iter().map(|d| Binner::new(d, K)).collect();
    let wbs: Vec<Binner> = inp.w_dims.iter().map(|d| Binner::new(d, K)).collect();

    let mut samples: Vec<Sample> = Vec::with_capacity(n);
    let mut zw: Vec<String> = Vec::with_capacity(n);
    for i in 0..n {
        let xb = bx.bin(inp.x[i]);
        let yb = by.bin(inp.y[i]);
        let zstr = join_bins(&zbs, &inp.z_dims, i);
        let wstr = join_bins(&wbs, &inp.w_dims, i);
        samples.push((xb, yb, zstr.clone()));
        zw.push(format!("{};{}", zstr, wstr)); // re-conditioning key = (Z, W)
    }

    let misspec: Box<dyn Fn(&[Sample]) -> Vec<Sample>> = {
        let zw = zw.clone();
        Box::new(move |s: &[Sample]| {
            s.iter().enumerate().map(|(i, (x, y, _z))| (*x, *y, zw[i].clone())).collect()
        })
    };

    let r = audit(&samples, inp.reps, inp.seed, 4.0, 0.005, std::slice::from_ref(&misspec));

    let verdict = if !inp.identifiable {
        CouplingVerdict::Unidentifiable
    } else {
        match r.decision {
            ChannelDecision::ConsistentWithNull => CouplingVerdict::AirGapHeld,
            ChannelDecision::ResidualMisspecStable => CouplingVerdict::ObserverContamination,
            ChannelDecision::ResidualMisspecFragile => CouplingVerdict::ConfoundedArtifact,
            ChannelDecision::ResidualDependence => CouplingVerdict::SuspectUnstressed,
        }
    };

    CouplingResult {
        name: inp.name.clone(),
        manifest_rule: inp.manifest_rule.clone(),
        verdict,
        mi: r.mi,
        cmi: r.cmi,
        null_mean: r.null_mean,
        z_score: r.z_score,
        identifiable: inp.identifiable,
        note: inp.note.clone(),
    }
}
